homeassistant/components/weather/__init__.py

```diff
@@ -1075,3 +1099,39 @@ async def async_update_listeners(
             converted_forecast_list = self._convert_forecast(native_forecast_list)
             for listener in self._forecast_listeners[forecast_type]:
                 listener(converted_forecast_list)
+
+
+def raise_unsupported_forecast(entity_id: str, forecast_type: str) -> None:
+    """Raise error on attempt to get an unsupported forecast."""
+    raise HomeAssistantError(
+        f"Weather entity '{entity_id}' does not support '{forecast_type}' forecast"
+    )
+
+
+async def async_get_forecast_service(
+    weather: WeatherEntity, service_call: ServiceCall
+) -> ServiceResponse:
+    """Get weather forecast."""
+    forecast_type = service_call.data["type"]
+    supported_features = weather.supported_features or 0
+    if forecast_type == "daily":
+        if (supported_features & WeatherEntityFeature.FORECAST_DAILY) == 0:
+            raise_unsupported_forecast(weather.entity_id, forecast_type)
+        native_forecast_list = await weather.async_forecast_daily()
+    elif forecast_type == "hourly":
+        if (supported_features & WeatherEntityFeature.FORECAST_HOURLY) == 0:
+            raise_unsupported_forecast(weather.entity_id, forecast_type)
+        native_forecast_list = await weather.async_forecast_hourly()
+    else:
+        if (supported_features & WeatherEntityFeature.FORECAST_TWICE_DAILY) == 0:
+            raise_unsupported_forecast(weather.entity_id, forecast_type)
+        native_forecast_list = await weather.async_forecast_twice_daily()
+    if native_forecast_list is None:
```

Comment:
Should this not raise instead of a forecast is missing?

Reply:
We allow a weather entity so support a forecast, but return `None` if no data is available. I'm not sure we should raise instead?